### PR TITLE
Separate the TCP and terse JSON formatting code.

### PR DESCRIPTION
**Pull request by clokep**
This does not change any functionality but breaks the dependency between the Twisted TCP logging code and it having to do the terse JSON output. The goal was a better separation between "transport" and "formatter". I'm hoping this will help make the code a bit more readable and help with being able to enable structured logging at the same time as standard logging (although this PR does not allow that).

It is pretty much:

* Copying and pasting `LogProducer` and `TerseJSONToTCPLogObserver` (renamed to `TCPLogObserver`) from `synapse.logging._terse_json` to `synapse.logging._remote`.
* Add a wrapper to `synapse.logging._terse_json` to hook up `TCPLogObserver` with the terse JSON formatting code.
* Pass throw the `format_event` callable through `TCPLogObserver` into `LogProducer` instead of a metadata dictionary.

This is working towards #8588.

### CODE REVIEW

File: tests/logging/test_terse_json.py [hunk 1]
```diff
@@ -78,7 +78,7 @@ def test_log_output(self):
             "server_name",
             "name",
         ]
-        self.assertEqual(set(log.keys()), set(expected_log_keys))
```
Comment (erikjohnston) [inline tests/logging/test_terse_json.py:81]: Why this change?

**Review by erikjohnston — APPROVED** [review]
Other than the question I have about changing the test,  I think it works

File: tests/logging/test_terse_json.py (hunk 1 above)
Comment (clokep) [reply tests/logging/test_terse_json.py:81]: While debugging the code this test was failing, but using `assertCountEquals` gives a better error message. 🤷 I can back it out if you want.

Comment (erikjohnston) [reply tests/logging/test_terse_json.py:81]: Ah right. It's just the new test is weaker, so I'd prefer we kept it as it is if possible.

Comment (clokep) [reply tests/logging/test_terse_json.py:81]: `assertCountEqual` should be a *stronger* test:

> Test that sequence *first* contains the same elements as *second*, regardless of their order. When they don’t, an error message listing the differences between the sequences will be generated.
>
> Duplicate elements are *not* ignored when comparing *first* and *second*. It verifies whether each element has the same count in both sequences. Equivalent to: `assertEqual(Counter(list(first)), Counter(list(second)))` but works with sequences of unhashable objects as well.

Casting to a `set` first means that duplicates are allowed.

How do you see it being a weaker test?

Comment (erikjohnston) [reply tests/logging/test_terse_json.py:81]: Oooooooooooooh, sorry, for some reason I was sure it was just testing the lengths matched. Ignore me
